csrc/id_model/id_model.cpp

```diff
@@ -627,10 +627,29 @@ std::unordered_map<ValGroup, IterDomain*> IdModel::buildLoopPromotionMap(
           idGraph(IdMappingMode::LOOP),
           inlining_info);
 
+  // At this point, most of loop groups should have correct promoted
+  // IDs. However, non-inlined loop groups may miss promotion that
+  // should be propagated from parent ID groups, e.g., iS50 of T2 in
+  // Indexing19. Its parent ID loop group is promoted, but the loop
+  // group of iS50 is not found yet.
+
+  // Step 4: In order to fully propagate the loop graph promotions, first
+  // propagate them to the IEL groups, which are then used to
+  // propagate back to the loop groups in Step 5. Unlike Step 2, the
+  // initial IEL promotion map is empty and is populated with the loop
+  // promotion map as we traverse down the IEL graph.
+  std::unordered_map<ValGroup, IterDomain*> final_iel_promotion_map;
+  propagatePromotionsInIELGraph(
+      iel_graph,
+      final_iel_promotion_map,
+      idGraph(IdMappingMode::LOOP),
+      loop_promotion_map,
+      true);
```

Comment:
Does it make sense to do the following instead?:
```C++
for (auto entry : iel_promotion_map) {
  ValGroup iel_group = entry->first;
  ValGroup loop_group = idGraph(IdMappingMode::LOOP).toGroup(iel_group->front());
  auto it = loop_promotion_map.find(loop_group);
  if (it != loop_promotion_map.end()) {
    entry->second = it->second;
  }
}
propagatePromotionsInIELGraph(iel_graph, iel_promotion_map, require_loop_mapped_promotion=true);
```
I think my biggest problem with `IdModel` is this is so complicated that I can not fit it into my mind. IIUC, changing it to the above code is equivalent to the current approach, but the mental model will be easier.

Reply:
This is actually related to https://github.com/NVIDIA/Fuser/pull/2003#discussion_r1548658814.

Using the Step 3 results needs to consider the condition checked by `hasUniqueInputLoopGraphs`, so the suggested code would result in the double propagation. We could avoid that by selectively updating `iel_promotion_map`, but that would mean we would need to look at inputs and outputs and propagate `loop_promotion_map` only to outputs in some cases. I'd say that would be almost equally complicated as the current version.

Reply:
I see, thanks for the explanation. Could you open an issue for https://github.com/NVIDIA/Fuser/pull/2003#discussion_r1548658814?

Reply:
> Could you open an issue for https://github.com/NVIDIA/Fuser/pull/2003#discussion_r1548658814?

What issue are you referring to? The broadcast forwarding?

Reply:
Yes

Reply:
Added: #2030